### ISSUE TITLE
How to evaluate YOLOv8 TensorFlow Lite (TFLite) inference on Coral

### ISSUE THREAD

**Issue by AlaaArboun** [ultralytics/ultralytics#8649]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

```bash
import cv2
import numpy as np
from tkinter import filedialog
import json
import os
from tflite_runtime.interpreter import Interpreter, load_delegate
import threading
from queue import Queue
import time
from sklearn.metrics.pairwise import euclidean_distances
import tkinter as tk 



class ObjectDetector:
    def __init__(self, model_path, label_path, threshold=0.3):
        self.threshold = threshold
        self.interpreter = Interpreter(model_path=model_path, experimental_delegates=[load_delegate('edgetpu.dll')])
        self.interpreter.allocate_tensors()
        self.input_details = self.interpreter.get_input_details()
        self.output_details = self.interpreter.get_output_details()
        self.input_shape = self.input_details[0]['shape'][1:]
        self.class_labels = self.load_class_labels(label_path)

        self.frame_count = 0
        self.start_time = time.time()


    def load_class_labels(self, label_path):
        with open(label_path, 'r') as json_file:
            class_labels = json.load(json_file)
        return class_labels

    def preprocess_frame(self, frame):
        image_resized = cv2.resize(frame, (self.input_shape[0], self.input_shape[1]))
        image_float32 = image_resized.astype(np.float32)
        image_normalized = image_float32 / 255.0
        image_with_batch = np.expand_dims(image_normalized, 0)
        return image_with_batch

    def nms(self, boxes, scores, threshold):
        if len(boxes) == 0:
            return []
        x1 = boxes[:, 0]
        y1 = boxes[:, 1]
        x2 = boxes[:, 2]
        y2 = boxes[:, 3]
        areas = (x2 - x1 + 1) * (y2 - y1 + 1)
        order = scores.argsort()[::-1]
        keep = []
        while order.size > 0:
            i = order[0]
            keep.append(i)
            xx1 = np.maximum(x1[i], x1[order[1:]])
            yy1 = np.maximum(y1[i], y1[order[1:]])
            xx2 = np.minimum(x2[i], x2[order[1:]])
            yy2 = np.minimum(y2[i], y2[order[1:]])
            w = np.maximum(0.0, xx2 - xx1 + 1)
            h = np.maximum(0.0, yy2 - yy1 + 1)
            overlap = (w * h) / areas[order[1:]]
            inds = np.where(overlap <= threshold)[0]
            order = order[inds + 1]
        return keep
    
    def perform_inference(self, frame, preprocessed_frame, orig_h, orig_w):
        self.interpreter.set_tensor(self.input_details[0]['index'], preprocessed_frame)
        self.interpreter.invoke()
        output_data = self.interpreter.get_tensor(self.output_details[0]['index'])
        output = np.copy(output_data[0])

        boxes = []
        scores = []



        for i in range(output.shape[1]):
            detection = output[:, i]
            x_center, y_center, width, height = detection[:4]
            x1 = max(0, int((x_center - width / 2) * orig_w))
            y1 = max(0, int((y_center - height / 2) * orig_h))
            x2 = min(orig_w - 1, int((x_center + width / 2) * orig_w))
            y2 = min(orig_h - 1, int((y_center + height / 2) * orig_h))
            score = np.max(detection[4:])
            boxes.append([x1, y1, x2, y2])
            scores.append(score)

        boxes = np.array(boxes)
        scores = np.array(scores)
        keep = self.nms(boxes, scores, 0.7)

        for idx in keep:
            x1, y1, x2, y2 = boxes[idx]
            score = scores[idx]
            detection = output[:, idx]
            cls = np.argmax(output[:, idx][4:])
            if score >= self.threshold:
                class_name = self.class_labels[str(cls)]
                cv2.rectangle(frame, (x1, y1), (x2, y2), (0, 0, 255), thickness=2)
                cv2.putText(frame, f"{class_name}: {score:.2f}", (x1, y1 - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 2)

        self.frame_count += 1

    def get_fps(self):
        current_time = time.time()
        elapsed_time = current_time - self.start_time
        if elapsed_time > 0:
            fps = self.frame_count / elapsed_time
            return fps
        else:
            return 0

    def detect_objects(self, frame):
        preprocessed_frame = self.preprocess_frame(frame)
        orig_h, orig_w = frame.shape[:2]
        self.perform_inference(frame, preprocessed_frame, orig_h, orig_w)

    def run_detection(self, cap, output_queue):
        while cap.isOpened():
            ret, frame = cap.read()
            if not ret:
                break
            output_queue.put(frame)

        cap.release()
        output_queue.put(None)

class App:
    def __init__(self, root):
        self.root = root
        self.root.title("Object Detection")
        self.detector = None

        select_button = tk.Button(root, text="Select Video", command=self.open_video)
        select_button.pack()

    def open_video(self):
        file_path = filedialog.askopenfilename()
        if file_path:
            cap = cv2.VideoCapture(file_path)
            output_queue = Queue()
            self.detector = ObjectDetector(
                model_path='yolov8m_saved_model\yolov8m_integer_quant.tflite',
                label_path='C:/object-detection-coral/label_files/labels_coco.json'
            )
            detection_thread = threading.Thread(target=self.detector.run_detection, args=(cap, output_queue))
            detection_thread.start()
            self.process_output(output_queue)

    def process_output(self, output_queue):
        while True:
            frame = output_queue.get()
            if frame is None:
                break
            self.detector.detect_objects(frame)
            cv2.putText(frame, f"FPS: {self.detector.get_fps():.2f}", (10, 30), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 255, 0), 2)
            cv2.imshow('Object Detection', frame)
            if cv2.waitKey(1) & 0xFF == ord('q'):
                break

if __name__ == "__main__":
    root = tk.Tk()
    app = App(root)
    root.mainloop()
```

**Comment by glenn-jocher** [comment]
@AlaaArboun hey there! 👋 It looks like you're diving into running YOLOv8 TensorFlow Lite (TFLite) inference on Coral devices. Your script seems well-structured for object detection using TFLite on a Coral device. However, to specifically evaluate the performance or accuracy of your YOLOv8 TFLite model on Coral, you might consider the following steps:

1. **Accuracy Evaluation**: To evaluate the model's accuracy, you'll need a labeled dataset for comparison. You can run your model on this dataset and compare the predictions against the true labels. Metrics such as precision, recall, and mAP (mean Average Precision) are commonly used for this purpose.

2. **Performance Metrics**: For performance, you're already calculating FPS (Frames Per Second) which is great for real-time inference scenarios. Additionally, consider measuring the inference time directly in your `perform_inference` method to get a more granular view of model performance.

3. **Visualization**: Your script includes visualization of detections which is excellent for qualitative evaluation. Ensure that the detections are accurate and consistent across frames for video inputs.

4. **Quantitative Analysis**: For a more quantitative analysis, consider logging detection results (coordinates, classes, confidence scores) to a file and using external tools or scripts to calculate accuracy metrics against a labeled test set.

5. **Optimization**: Since you're using Coral, ensure your model is fully optimized for Edge TPU using the [Edge TPU Compiler](https://coral.ai/docs/edgetpu/compiler/). This can significantly affect both accuracy and performance.

For a more detailed evaluation, you might need to adjust your script to output detection results in a format that can be easily compared with your ground truth labels. This could involve saving detection results to a JSON or CSV file during inference.

Remember, the key to effective model evaluation is having a reliable ground truth dataset and a clear understanding of the metrics that best represent success for your specific application. 🚀

Hope this helps! If you have further questions or run into issues, feel free to ask. Happy coding!
